packages/turbo-gen/src/utils/workspaceRoots.ts

```diff
@@ -0,0 +1,32 @@
+import path from "path";
+import type { Project } from "@turbo/workspaces";
+
+// This function is not perfect and could be improved to be more accurate.
+// Given a list of workspace globs, it aims to return a selectable list of paths that are valid workspace locations.
+// This current naive approach does not work with globs that contain nested wildcards, for example: `packages/*/utils` will not work.
+export function getWorkspaceRoots({
+  project,
+}: {
+  project: Project;
+}): Array<string> {
+  const allWorkspaces = project.workspaceData.workspaces;
+  const allWorkspacePaths = allWorkspaces.map((workspace) =>
+    path.relative(project.paths.root, workspace.paths.root)
+  );
+
+  // find valid workspace locations
+  const workspaceRoots = new Set<string>();
+  project.workspaceData.globs.forEach((glob) => {
+    if (allWorkspacePaths.includes(glob)) {
+      return;
+    } else if (glob.startsWith("!")) {
+      return;
+    } else {
+      const globParts = glob.split("/");
+      const globRoot = globParts[0];
+      workspaceRoots.add(globRoot);
+    }
+  });
+
+  return Array.from(workspaceRoots);
+}
```

Comment:
Is this logic something that could go in `@turbo/workspaces`?

Reply:
Yea, this, and "categorize workspaces" (bucket workspaces by their parent directory) is something I find myself doing a ton so I think we should move it there.